### PR TITLE
Add object detection with yolo

### PR DESCRIPTION
**Pull request by DominikKwiatkowski**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -128,3 +128,11 @@ dmypy.json
 # Pyre type checker
 .pyre/
 .idea/
+
+# Photos
+*.jpg
+*.png
+
+# default yolo models
```
Comment (Qubaef) [inline .gitignore]: "YOLO model directory" would probably be more precise.

File: drones/image_processing/image_processing.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+from drones.image_processing.utils import distance_to_camera, vector_to_centre
+
+
+class ImageProcessing:
+    def __init__(self):
+        self.yolo = YoloDetection()
+
+    def process_image(self, image: np.ndarray) -> list:
+        """
+        Detect object, count its distance from camera, pitch and yaw
```
Comment (Qubaef) [inline drones/image_processing/image_processing.py]: Add period at the end of a sentence.

File: drones/image_processing/image_processing.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        Parameters:
+        ----------
+        image: np.ndarray
+            image to be processed
+
+        Returns:
+        ----------
+            list of directions and distances of all objects: list and typing.Tuple[float,float,float] inside
```
Comment (Qubaef) [inline drones/image_processing/image_processing.py]: Fix gramar and add period at the end of a docstring.

File: drones/image_processing/image_processing.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        result_list = []
+
+        if len(detection_list) > 0:
+            for detection in detection_list:
+                x, y, width = detection
+                distance = distance_to_camera(real_width, focal, width)
+                vector = vector_to_centre(image_width, image_height, (x, y), 0.5)
+                result_list.append(
+                    (
+                        -(vector[0] / image_width * float(config["FIELD_OF_VIEW"])),
```
Comment (Qubaef) [inline drones/image_processing/image_processing.py]: Remove redundant parenthesis.

File: drones/image_processing/yolo.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+from yolov5.models.experimental import attempt_load
+from yolov5.utils.datasets import letterbox
+from yolov5.utils.general import check_img_size, non_max_suppression, scale_coords, xyxy2xywh
+from yolov5.utils.torch_utils import select_device
+import logging
+
+
+class YoloDetection:
+    def __init__(self):
+        # initialize config
```
Comment (Qubaef) [inline drones/image_processing/yolo.py]: If you start comments with a capital letter, make them all that way. Add period at the end of a sentence.

File: drones/image_processing/yolo.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+        self.device_type = self.config["DEVICE"]
+        self.model = attempt_load(self.config["NETWORK_PATH"], map_location=self.device_type)  # load FP32 model
+        self.stride = int(self.model.stride.max())  # model stride
+        self.conf_tres = float(self.config["CONFIDENCE_THRESHOLD"])
+        self.img_size = int(self.config["IMG_SIZE"])
+        self.iou_thres = float(self.config["IOU_THRESHOLD"])
+
+    def detect(self, img0: np.ndarray) -> list:
+        """Detect object on image using provided weights. Objects are detected by YOLO neural network
```
Comment (Qubaef) [inline drones/image_processing/yolo.py]: Period at the end of a sentence.

File: drones/image_processing/image_processing.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        config = config_parser["OBJECT"]
+
+        focal = int(config["FOCAL"])
+        real_width = float(config["WIDTH"])
+        detection_list = self.yolo.detect_object_yolo(image)
+        image_width = image.shape[1]
+        image_height = image.shape[0]
+        result_list = []
+
+        if len(detection_list) > 0:
```
Comment (Qubaef) [inline drones/image_processing/image_processing.py]: According to PEP 8 style guide, emptiness of the list should be verified by `if detection_list:`

File: drones/image_processing/yolo.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+        Parameters:
+        ----------
+            img0: np.ndarray
+                image on which objects detection will be proceeded
+        Returns:
+        ----------
+            It returns list of all detected objects on photo. Every element consist 5 values. First one is recognized
```
Comment (Qubaef) [inline drones/image_processing/yolo.py]: Fix grammar ("on the ~photo~ image", "consist of",  space before parenthesis).

File: drones/image_processing/yolo.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+                # Rescale boxes from img_size to im0 size
+                det[:, :4] = scale_coords(image.shape[2:], det[:, :4], img0.shape).round()
+
+                # Write results
+                for *xyxy, conf, cls in reversed(det):
+                    xywh = (xyxy2xywh(torch.tensor(xyxy).view(1, 4)) / gn).view(-1).tolist()  # normalized xywh
+                    line = (names[int(cls)], *xywh)  # label format
+                    result.append(line)
+
+        logging.getLogger(__name__).info(f"Processed image, processing time: ({time.time() - t0:.3f}s)")
```
Comment (Qubaef) [inline drones/image_processing/yolo.py]: You can add debug log messages, containing positions of all detected objects (either here, or in `detect_object_yolo` below).

File: drones/image_processing/yolo.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+        Parameters:
+        ----------
+            img0: np.ndarray
+                image on which objects detection will be proceeded
+
+        Returns:
+        ----------
+            center_and_diameter: list of tuples(int, int, int)
+                Every found instance of object defined in config file has it's own tuple. Every tuple have 3 values,
```
Comment (Qubaef) [inline drones/image_processing/yolo.py]: Fix grammar and formatting ("has 3 values", "which represents", space before parenthesis).

File: drones/image_processing/yolo.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+        img = img[:, :, ::-1].transpose(2, 0, 1)  # BGR to RGB, to 3x416x416
+        img = np.ascontiguousarray(img)
+
+        # Get names and colors
+        names = self.model.module.names if hasattr(self.model, "module") else self.model.names
+
+        device = select_device(self.device_type)
+
+        # Run inference
+        if device.type != "cpu":
```
Comment (Qubaef) [inline drones/image_processing/yolo.py:71]: What if device.type is not "cpu"? If we assume that model should not run, this config parameter is not required.

File: drones/image_processing/yolo.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        """
+
+        # Using yolo detect object on photo
+        results = self.detect(image)
+
+        image_width = image.shape[1]
+        image_height = image.shape[0]
+
+        result_list = []
+        # Find proper class on photo. Should be only one represent of this class
```
Comment (Qubaef) [inline drones/image_processing/yolo.py]: I don't get this comment, but I guess it's not right, as we are collecting all instances of the class. Could you explain what you meant?

**Review by Qubaef — CHANGES_REQUESTED** [review]
In addition to all the comments, I suspect that the test.jpg file should be excluded from the commit because it is the .jpg image you tried to ignore.

File: drones/image_processing/image_processing.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+import drones.image_processing.normalization as normalization
+from drones.image_processing.yolo import YoloDetection
+from drones.image_processing.utils import distance_to_camera, vector_to_centre
+
+
+class ImageProcessing:
+    def __init__(self):
+        self.yolo = YoloDetection()
+
+    def process_image(self, image: np.ndarray) -> list:
```
Comment (bazylip) [inline drones/image_processing/image_processing.py]: Use `typing.List`.

File: drones/image_processing/image_processing.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+import cv2 as cv
+import imutils
+import configparser
+import typing
+import drones.image_processing.normalization as normalization
+from drones.image_processing.yolo import YoloDetection
+from drones.image_processing.utils import distance_to_camera, vector_to_centre
+
+
+class ImageProcessing:
```
Comment (bazylip) [inline drones/image_processing/image_processing.py:12]: Missing class level docstring.

File: drones/image_processing/yolo.py [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+            self.classes = None  # type: ignore
+
+        self.device_type = self.config["DEVICE"]
+        self.model = attempt_load(self.config["NETWORK_PATH"], map_location=self.device_type)  # load FP32 model
+        self.stride = int(self.model.stride.max())  # model stride
+        self.conf_tres = float(self.config["CONFIDENCE_THRESHOLD"])
+        self.img_size = int(self.config["IMG_SIZE"])
+        self.iou_thres = float(self.config["IOU_THRESHOLD"])
+
+    def detect(self, img0: np.ndarray) -> list:
```
Comment (bazylip) [inline drones/image_processing/yolo.py]: Use `typing.List`.

File: drones/image_processing/image_processing.py [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        focal = int(config["FOCAL"])
+        real_width = float(config["WIDTH"])
+        detection_list = self.yolo.detect_object_yolo(image)
+        image_width = image.shape[1]
+        image_height = image.shape[0]
+        result_list = []
+
+        if len(detection_list) > 0:
+            for detection in detection_list:
+                x, y, width = detection
```
Comment (bazylip) [inline drones/image_processing/image_processing.py]: Why not unpack `detection` directly in `for` loop?

File: drones/image_processing/yolo.py [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+                # Write results
+                for *xyxy, conf, cls in reversed(det):
+                    xywh = (xyxy2xywh(torch.tensor(xyxy).view(1, 4)) / gn).view(-1).tolist()  # normalized xywh
+                    line = (names[int(cls)], *xywh)  # label format
+                    result.append(line)
+
+        logging.getLogger(__name__).info(f"Processed image, processing time: ({time.time() - t0:.3f}s)")
+        return result
+
+    def detect_object_yolo(self, image: np.ndarray) -> list:
```
Comment (bazylip) [inline drones/image_processing/yolo.py]: Use `typing.List`.

File: drones/image_processing/yolo.py [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+        """Function will detect objects on given image and return position and width of object, which class is chosen
```
Comment (bazylip) [inline drones/image_processing/yolo.py]: Docstring summary is too complex. Use summary, newline and further description.

File: drones/image_processing/yolo.py [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+
+        # Inference
+        pred = self.model(image)[0]
+
+        # Apply NMS
+        pred = non_max_suppression(pred, self.conf_tres, self.iou_thres, classes=self.classes)
+
+        # Process detections
+        for i, det in enumerate(pred):  # detections per image
+            gn = torch.tensor(img0.shape)[[1, 0, 1, 0]]  # normalization gain whwh
```
Comment (bazylip) [inline drones/image_processing/yolo.py]: What does `whwh` stand for?

File: drones/image_processing/yolo.py (hunk 9 above)
Comment (bazylip) [inline drones/image_processing/yolo.py]: Create logger instance at file level instead of locally in methods.

File: requirements.txt [hunk 20]
```diff
@@ -3,3 +3,4 @@ numpy==1.20.2
 opencv-python==4.5.1.48
 pre-commit~=2.11.0
 pytest==6.2.2
+yolov5
```
Comment (bazylip) [inline requirements.txt]: Specify version of library (at least minimal).

File: drones/image_processing/yolo.py [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+                image on which objects detection will be proceeded
+
+        Returns:
+        ----------
+            center_and_diameter: list of tuples(int, int, int)
+                Every found instance of object defined in config file has it's own tuple. Every tuple have 3 values,
+                which represent object center coordinates(x,y) and width.
+                If the object is not detected list will be empty.
+        """
+
```
Comment (bazylip) [inline drones/image_processing/yolo.py]: Redundant empty line.

**Review by bazylip — CHANGES_REQUESTED** [review]
There is basically no logging. Suggestion to add more debug-level messages.

File: drones/image_processing/yolo.py (hunk 11 above)
Comment (DominikKwiatkowski) [reply drones/image_processing/yolo.py:71]: it is configuration in case of using GPU

File: drones/image_processing/yolo.py (hunk 10 above)
Comment (Qubaef) [reply drones/image_processing/yolo.py]: Still not resolved

File: drones/image_processing/yolo.py (hunk 9 above)
Comment (Qubaef) [reply drones/image_processing/yolo.py]: Still log messages missing

**Review by Qubaef — COMMENTED** [review]
Still .jpg file is included in the commit. It should not be there.

File: drones/image_processing/image_processing.py [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+    """
+
+    def __init__(self):
+        """
+        Creates instance of YOLO object, to perform detection using yolo. All parameters can be specified in config
+        file.
+        """
+        self.yolo = YoloDetection()
+
+    def process_image(self, image: np.ndarray) -> typing.List:
```
Comment (bazylip) [inline drones/image_processing/image_processing.py]: Please specify what type of elements will be in return `typing.List`.

File: drones/image_processing/yolo.py (hunk 10 above)
Comment (Qubaef) [reply drones/image_processing/yolo.py]: Still not resolved, but idc anymore

File: drones/image_processing/yolo.py [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+        # Parse classes from config.
+        config_classes = self.config["CLASSES"]
+        if config_classes != "None":
+            det_classes = config_classes.split(" ")
+            self.classes = []
+            for det_class in det_classes:
+                self.classes.append(int(det_class))
+        else:
+            self.classes = None  # type: ignore
```
Comment (bazylip) [inline drones/image_processing/yolo.py]: Why do you use `type: ignore` here?

File: drones/image_processing/yolo.py [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+
+        self.device_type = self.config["DEVICE"]
+        self.model = attempt_load(self.config["NETWORK_PATH"], map_location=self.device_type)  # load FP32 model
+        self.stride = int(self.model.stride.max())  # model stride
+        self.conf_tres = float(self.config["CONFIDENCE_THRESHOLD"])
+        self.img_size = int(self.config["IMG_SIZE"])
+        self.iou_thres = float(self.config["IOU_THRESHOLD"])
+
+    def detect(self, img0: np.ndarray) -> typing.List:
```
Comment (bazylip) [inline drones/image_processing/yolo.py]: Please specify `typing.List` members.

File: drones/image_processing/yolo.py [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+                    line = (names[int(cls)], *xywh)  # label format
+                    result.append(line)
+                    self.log.info(
+                        f"found class {line[0]} in position {line[1]}, {line[2]} of size {line[3]}, {line[4]}"
+                    )
+
+        self.log.info(f"Processed image, processing time: ({time.time() - t0:.3f}s)")
+        return result
+
+    def detect_object_yolo(self, image: np.ndarray) -> typing.List:
```
Comment (bazylip) [inline drones/image_processing/yolo.py]: Please specify List members.

**Review by bazylip — CHANGES_REQUESTED** [review]
All these changes require unit tests but this is your organisatonial decision whether to include them or not.

**Review by Qubaef — APPROVED** [review]
Looks fine now

**Review by bazylip — APPROVED** [review]

**Pull request merged**